### PR TITLE
HW ex4

### PR DESCRIPTION
**Pull request by barakmaron**

### CODE REVIEW

File: src/ex4/dist/clients/ApiRequest.mjs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+
```
Comment (or-nuri-monday) [inline src/ex4/dist/clients/ApiRequest.mjs:1]: You could have generalized it even more. We can talk about it if you want.

File: src/ex4/dist/main.js [hunk 2]
```diff
@@ -1,52 +1,194 @@
...
-            const listItem = document.createElement("li");
-            listItem.classList.add('list-item');
-            listItem.innerHTML = item;
+  /**
+     * add button call back function
+     */
+  async AddButtonCallBack() {
+    const task_text_from_user = this.dom_manager.task_input.value;    
+    // check if input is not empty
+    if (task_text_from_user) {
```
Comment (or-nuri-monday) [inline src/ex4/dist/main.js:48]: It is a best practice to check if the `task_text_from_user` is different from an empty string. You could also trim the value.

File: src/ex4/dist/main.js [hunk 3]
```diff
@@ -1,52 +1,194 @@
...
+      this.dom_manager.CreateLoader();
+      // clear error message
+      this.dom_manager.ClearErrorEmptyTask();
+      try {
+        // send the task to item manager      
+        const add_promise = Promise.resolve(AddNewResourceRequest('task', { task: task_text_from_user }));
+        // wait to get response before re-rendering
+        const add_response = await add_promise;
+        // status 201 is for created
+        if (add_response.status !== 201)
```
Comment (or-nuri-monday) [inline src/ex4/dist/main.js:58]: You are already checking that the status is 201 in `AddNewResourceRequest`

File: src/ex4/dist/services/DomManager.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  }
+
+  CreateElementFromHtml(html) {
+    const temp = document.createElement('template');
+    const html_parsed = html.trim();
+    temp.innerHTML = html_parsed;
+    return temp.content.firstChild;
+  }
+
+  CreateLoader()
```
Comment (or-nuri-monday) [inline src/ex4/dist/services/DomManager.js:148]: ShowLoader and HideLoader would be better names

File: src/ex4/server/middleware/ErrorHandler.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+export default function ErrorHandler(err, req, res, next) {
+    if (err.length) {
+        err.forEach(error => {
+            console.log("Received error", error.message);
```
Comment (or-nuri-monday) [inline src/ex4/server/middleware/ErrorHandler.js:4]: Duplicated code

File: src/ex4/server/controllers/TaskController.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import ItemManagerService from "../services/ItemManager.mjs";
```
[the remark this answers is not in the document]
Comment (barakmaron) [reply src/ex4/server/controllers/TaskController.js:1]: didn't understand what you meant, can you be more specific?

File: src/ex4/dist/clients/ApiRequest.mjs (hunk 1 above)
Comment (barakmaron) [inline src/ex4/dist/clients/ApiRequest.mjs:1]:
```suggestion
export async function SendRequest(url, action = "GET", body = null) {
  try {
    const response = await fetch(`/${url}`, {
      method: action,
      headers: { 'Content-Type': 'application/json' },
      body: body || ""
    });
    if (!response.ok) {
      throw new Error(`Cant ${action} ${url}`);
    }
    return await response.json();
  }
  catch (error) {
    throw (error);
  }
}

```

I thought about this function @or-nuri-monday

**Pull request merged**